src/main/java/org/mastodon/mamut/feature/branch/BranchNSubBranchSpotsFeatureComputer.java

```diff
@@ -0,0 +1,92 @@
+package org.mastodon.mamut.feature.branch;
+
+import org.mastodon.collection.RefSet;
+import org.mastodon.graph.algorithm.RootFinder;
+import org.mastodon.mamut.feature.MamutFeatureComputer;
+import org.mastodon.mamut.model.branch.BranchLink;
+import org.mastodon.mamut.model.branch.BranchSpot;
+import org.mastodon.mamut.model.branch.ModelBranchGraph;
+import org.mastodon.properties.IntPropertyMap;
+import org.scijava.ItemIO;
+import org.scijava.plugin.Parameter;
+import org.scijava.plugin.Plugin;
+
+/**
+ * Computes the total number of successors of a branch spot in the whole track
+ * sub-tree of this branch spot.
+ *
+ * In the following example this number would equal to following branchSpots as
+ * follows:
+ * 
+ * <pre>
+ *                         branchSpot0
+ *  	       ┌──────────────┴─────────────────┐
+ *  	       │                                │
+ *  	   branchspot1                      branchSpot2
+ *  	┌──────┴───────┐
+ *  	│              │
+ *  branchspot3 branchSpot4
+ * </pre>
+ * 
+ * <ul>
+ * <li>{@code branchSpot0 = 4}</li>
+ * <li>{@code branchSpot1 = 2}</li>
+ * <li>{@code branchSpot2 = 0}</li>
+ * <li>{@code branchSpot3 = 0}</li>
+ * <li>{@code branchSpot4 = 0}</li>
+ * </ul>
+ */
+@Plugin( type = MamutFeatureComputer.class )
+public class BranchNSubBranchSpotsFeatureComputer implements MamutFeatureComputer
+{
+	@Parameter
+	private ModelBranchGraph branchGraph;
+
+	@Parameter( type = ItemIO.OUTPUT )
+	private BranchNSubBranchSpotsFeature output;
+
+	@Override
+	public void createOutput()
+	{
+		if ( null == output )
+			output = new BranchNSubBranchSpotsFeature( new IntPropertyMap<>( branchGraph.vertices().getRefPool(), -1 ) );
+	}
+
+	@Override
+	public void run()
+	{
+		computeNumberOfSubtreeNodes();
+	}
+
+	private void computeNumberOfSubtreeNodes()
+	{
+		final RefSet< BranchSpot > roots = RootFinder.getRoots( branchGraph );
+		for ( final BranchSpot root : roots )
+		{
+			numberOfNodes( root, null, output );
+		}
+		output.map.getMap().forEachKey( object -> {
+			output.map.set( object, output.map.get( object ) - 1 );
+			return true;
+		} );
+	}
+
+	private static void numberOfNodes(BranchSpot branchSpot1, BranchSpot branchSpot2, BranchNSubBranchSpotsFeature output)
+	{
+		output.map.set( branchSpot1, 1 );
+		for ( BranchLink branchLink: branchSpot1.outgoingEdges() )
+		{
+			BranchSpot childBranchSpot = branchLink.getTarget();
+
+			// condition to omit reverse path from children to parent
+			if(childBranchSpot.equals( branchSpot2 ))
+				continue;
```

Comment:
This is a test to prevent and infinite iteration on loops. But it only works for very specific loops.
An easy strategy to prevent iteration in a loop would be to make sure that you only follow an edge "source -> target" if a condition target.getTimepoint() > source.getTimepoint() holds.

There are also multiple classes to aimed at helping to perform a depth first iteration:
* https://github.com/mastodon-sc/mastodon-graph/blob/master/src/main/java/org/mastodon/graph/algorithm/traversal/DepthFirstSearch.java (automatically detects loops)
* https://github.com/mastodon-sc/mastodon/blob/master/src/main/java/org/mastodon/util/DepthFirstIteration.java (loop detection must be done by the client code)

Reply:
Could you provide a test case, where the current code would break?
Otherwise, I would leave the logic as it is, since the unit-tests show that it works.

Reply:
The DepthFirstSearch class should be used to implement this method see suggested change for `BranchNLeavesComputer`

Reply:
Ok, the code now uses the DepthFirstSearch class